src/server/frontend_wayland/foreign_toplevel_manager_v1.cpp

```diff
@@ -458,19 +482,156 @@ void mf::ForeignSurfaceObserver::renamed(ms::Surface const*, std::string const&
 }
 
 void mf::ForeignSurfaceObserver::application_id_set_to(
-    scene::Surface const*,
+    scene::Surface const* surface,
     std::string const& application_id)
 {
     std::lock_guard lock{mutex};
 
     std::string id = application_id;
-    with_toplevel_handle(lock, [id](ForeignToplevelHandleV1& handle)
+    with_toplevel_handle(lock, [&](ForeignToplevelHandleV1& handle)
         {
-            handle.send_app_id_event(id);
-            handle.send_done_event();
+            auto app_id = desktop_file_manager->resolve_app_id(surface);
+            if (!app_id.empty())
+            {
+                handle.send_app_id_event(app_id);
+                handle.send_done_event();
+            }
         });
 }
 
+// GDesktopFileCache
+mf::GDesktopFileCache::GDesktopFileCache(const std::shared_ptr<MainLoop> &main_loop)
+    : main_loop{main_loop},
+      inotify_fd{inotify_init()}
+{
+    refresh_app_cache();
+
+    if (inotify_fd < 0)
+    {
+        mir::log_error("Unable to initialize inotify. Automatic desktop app refresh will NOT happen.");
+        return;
+    }
+
+    // Set up a watch on the data directories
+    auto data_directories = g_get_system_data_dirs();
+    for (int i = 0; data_directories[i] != NULL; i++)
+    {
+        auto application_directory = std::string(data_directories[i]) + "/applications";
+        int config_path_wd = inotify_add_watch(
+            inotify_fd,
+            application_directory.c_str(),
+            IN_CREATE | IN_DELETE | IN_CLOSE_WRITE | IN_MOVED_TO | IN_MOVED_FROM);
+
+        if (config_path_wd < 0)
+        {
+            mir::log_error("Unable to watch directory %s", application_directory.c_str());
+            continue;
+        }
+
+        config_path_wd_list.push_back(config_path_wd);
+    }
+
+    main_loop->register_fd_handler({inotify_fd}, this, [this](int)
+    {
+        union
+        {
+            inotify_event event;
+            char buffer[sizeof(inotify_event) + NAME_MAX + 1];
+        } inotify_buffer;
+        if (read(inotify_fd, &inotify_buffer, sizeof(inotify_buffer)) < static_cast<ssize_t>(sizeof(inotify_event)))
+            return;
+
+        refresh_app_cache();
+    });
+}
+
+void mf::GDesktopFileCache::refresh_app_cache()
+{
+    auto free_app_infos = [](GList* app_infos)
+    {
+        g_list_free_full(app_infos, g_object_unref);
+    };
+    std::unique_ptr<GList, decltype(free_app_infos)> app_infos(g_app_info_get_all(), free_app_infos);
+    std::map<std::string, std::shared_ptr<DesktopFile>> new_id_to_app;
+    std::map<std::string, std::shared_ptr<DesktopFile>> new_wm_class_to_app_info;
+    std::map<std::string, std::shared_ptr<DesktopFile>> new_exec_to_app;
+
+    GList* info;
+    for (info = app_infos.get(); info != NULL; info = info->next)
+    {
+        GAppInfo *app_info = static_cast<GAppInfo*>(info->data);
+        if (!g_app_info_should_show(app_info))
```

Comment:
And the answer is: no, because there are cases where both a `NoDisplay` and should-display-in .desktop file will match, and we want to ensure that the normal desktop file wins.

cf: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/2721